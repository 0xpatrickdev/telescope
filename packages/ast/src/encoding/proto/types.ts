import * as t from '@babel/types';
import { getObjectName } from '@osmonauts/proto-parser';
import { pascal } from 'case';
import { identifier, tsPropertySignature, functionDeclaration } from '../../utils';
import { ProtoParseContext } from '../context';

export interface ProtoAny {
    type: string;
    name: string;
    [key: string]: any;
};
export interface ProtoEnum {
    name?: string;
    values: { [key: string]: number };
    comment?: string;
    comments?: { [key: string]: string };
}

export interface ProtoType {
    name?: string;
    oneofs?: { [key: string]: { oneof: string[], comment: string | undefined } },
    fields: {
        [key: string]: ProtoField;
    },
    comment: string | undefined;
}

export interface ProtoField {
    name?: string;
    scope?: string[];
    parsedType?: {
        name: string;
        type: string;
    },

    keyType?: string;
    rule?: string;
    type: string;
    id: number;
    options: {
        [key: string]: any;
        "(gogoproto.nullable)"?: boolean;
        "(cosmos_proto.scalar)"?: string;
        "(gogoproto.customtype)"?: string;
        "(gogoproto.moretags)"?: string;
        "(cosmos_proto.accepts_interface)"?: string;
    }
    comment?: string;
    import?: string;
    importedName?: string;
    scopeType?: string;
}

export const SCALAR_TYPES = [
    'string',
    'double',
    'float',
    'int32',
    'uint32',
    'sint32',
    'fixed32',
    'sfixed32',
    'int64',
    'uint64',
    'sint64',
    'fixed64',
    'sfixed64',
    'bytes',
    'bool'
];

export const NATIVE_TYPES = [
    ...SCALAR_TYPES,

    // TODO move these out
    'google.protobuf.Timestamp',
    'google.protobuf.Duration',
    'google.protobuf.Any',
]

export const getTSTypeFromProtoType = (type) => {
    switch (type) {
        case 'string':
            return t.tsStringKeyword();
        case 'double':
        case 'float':
        case 'int32':
        case 'uint32':
        case 'sint32':
        case 'fixed32':
        case 'sfixed32':
            return t.tsNumberKeyword();
        case 'int64':
        case 'uint64':
        case 'sint64':
        case 'fixed64':
        case 'sfixed64':
            return t.tsTypeReference(t.identifier('Long'))
        case 'bytes':
            return t.tsTypeReference(t.identifier('Uint8Array'));
        case 'bool':
            return t.tsBooleanKeyword();
        case 'google.protobuf.Timestamp':
            return t.tsTypeReference(t.identifier('Date'));
        case 'google.protobuf.Duration':
            return t.tsStringKeyword();
        case 'google.protobuf.Any':
            return t.tsTypeReference(t.identifier('Any'));
        default:
            throw new Error('getTSTypeFromProtoType() type not found');
    };
};

// https://github.com/protobufjs/protobuf.js/blob/master/src/types.js#L38-L54
export const types = {
    basic: {
        double: 1,
        float: 5,
        int32: 0,
        uint32: 0,
        sint32: 0,
        fixed32: 5,
        sfixed32: 5,
        int64: 0,
        uint64: 0,
        sint64: 0,
        fixed64: 1,
        sfixed64: 1,
        bool: 0,
        string: 2,
        bytes: 2
    },
    defaults: {
        double: 0,
        float: 0,
        int32: 0,
        uint32: 0,
        sint32: 0,
        fixed32: 0,
        sfixed32: 0,
        int64: 0,
        uint64: 0,
        sint64: 0,
        fixed64: 0,
        sfixed64: 0,
        bool: false,
        string: '',
        bytes: [],
        undefined: null
    },
    long: { int64: 0, uint64: 0, sint64: 0, fixed64: 1, sfixed64: 1 },
    mapKey: {
        int32: 0,
        uint32: 0,
        sint32: 0,
        fixed32: 5,
        sfixed32: 5,
        int64: 0,
        uint64: 0,
        sint64: 0,
        fixed64: 1,
        sfixed64: 1,
        bool: 0,
        string: 2
    },
    packed: {
        double: 1,
        float: 5,
        int32: 0,
        uint32: 0,
        sint32: 0,
        fixed32: 5,
        sfixed32: 5,
        int64: 0,
        uint64: 0,
        sint64: 0,
        fixed64: 1,
        sfixed64: 1,
        bool: 0
    }
};
export const getWireNumber = (type) => {
    if (types.basic.hasOwnProperty(type)) {
        return types.basic[type];
    }
    return 2;
};

export const getPackedWireNumber = (type) => {
    if (types.packed.hasOwnProperty(type)) {
        return types.packed[type];
    }
    return 2;
};

export const getTagNumber = (field: ProtoField) => {
    let wire = getWireNumber(field.type);
    if (field.rule === 'repeated') {
        wire = getPackedWireNumber(field.type);
    }
    if (field.parsedType.type === 'Enum') {
        wire = 0;
    }

    // for some reason I had to set wire = 2 for 
    // if (field.rule === 'repeated') wire = 2;
    // PinCodesProposal
    // maybe it's a big in ts-proto, and we're correct?

    return ((field.id << 3) | wire) >>> 0;
};


export const getDefaultTSTypeFromProtoType = (field: ProtoField) => {

    if (field.rule === 'repeated') {
        return t.arrayExpression([]);
    }

    if (field.keyType) {
        return t.objectExpression([])
    }

    switch (field.type) {
        case 'string':
            return t.stringLiteral('');
        case 'double':
        case 'float':
        case 'int32':
        case 'uint32':
        case 'sint32':
        case 'fixed32':
        case 'sfixed32':
            return t.numericLiteral(0);
        case 'uint64':
            return t.memberExpression(
                t.identifier('Long'),
                t.identifier('UZERO')
            );
        case 'int64':
        case 'sint64':
        case 'fixed64':
        case 'sfixed64':
            return t.memberExpression(
                t.identifier('Long'),
                t.identifier('ZERO')
            );
        case 'bytes':
            return t.newExpression(
                t.identifier('Uint8Array'),
                []
            );
        case 'bool':
            return t.booleanLiteral(false);

        // OTHER TYPES
        case 'google.protobuf.Timestamp':
            return t.identifier('undefined');
        case 'google.protobuf.Duration':
            return t.identifier('undefined');
        case 'google.protobuf.Any':
            return t.identifier('undefined');

        case 'cosmos.base.v1beta1.Coins':
            return t.arrayExpression([]);
        case 'cosmos.base.v1beta1.Coin':
            return t.identifier('undefined');

        default:
            // console.warn('getDefaultTSTypeFromProtoType() type not found: ' + type);
            return t.identifier('undefined');
    };
};

const lowerFirst = (str: string) => {
    return str.charAt(0).toLowerCase() + str.substring(1);
};
const upperFirst = (str: string) => {
    return str.charAt(0).toUpperCase() + str.substring(1);
};

export const getEnumToJsonName = (name) => {
    return lowerFirst(name) + 'ToJSON';
};

export const getEnumFromJsonName = (name) => {
    return lowerFirst(name) + 'FromJSON';
};

export const getFieldsTypeName = (field: ProtoField) => {
    if (field?.scope.length <= 1) return field.parsedType.name;
    const [_first, ...rest] = field.scope;
    return [...rest, field.parsedType.name].join('_');
};

export const getKeyTypeEntryName = (typeName: string, prop: string) => {
    return `${typeName}_${pascal(prop)}Entry`;
};

export const getBaseCreateTypeFuncName = (name) => {
    return `createBase${upperFirst(name)}`;
};

/*
  returns "Type | undefined"
*/
const optionalType = (type: t.TSType) => {
    return t.tsUnionType([
        type,
        t.tsUndefinedKeyword()
    ])
}

export const getOneOfs = (type: ProtoType) => {
    const keys = Object.keys(type.oneofs ?? {});
    if (!keys.length) return [];
    if (keys.length !== 1) throw new Error('getOneOfs() oneofs cardinality not known!');
    return type.oneofs[keys[0]].oneof;
};

export const getFieldOptionality = (field: ProtoField, isOneOf: boolean) => {
    return isOneOf || field?.options?.['(gogoproto.nullable)'];
};

const getProtoFieldTypeName = (field: ProtoField) => {
    let name = field.type;

    if (name.includes('.')) {
        const parts = name.split('.');
        name = parts[parts.length - 1];
    }

    if (!field.scope || !field.scope.length) return name;
    return getObjectName(name, field.scope);
};

const getProtoField = (field: ProtoField) => {
    let ast: any = null;
    let optional = false;

    if (field.options?.['(gogoproto.nullable)']) {
        optional = true;
    }

    if (NATIVE_TYPES.includes(field.type)) {
        ast = getTSTypeFromProtoType(field.type);
    } else {
        ast = t.tsTypeReference(t.identifier(getProtoFieldTypeName(field)));
    }

    if (field.rule === 'repeated') {
        ast = t.tsArrayType(ast);
    }

    if (field.keyType) {
        ast = t.tsUnionType([
            t.tsTypeLiteral([
                t.tsIndexSignature([
                    identifier('key',
                        t.tsTypeAnnotation(
                            getTSTypeFromProtoType(field.keyType)
                        )
                    )
                ],
                    t.tsTypeAnnotation(ast)
                )
            ])
        ]);
    }

    return ast;
};

export const createProtoType = (name: string, proto: ProtoType) => {
    const oneOfs = getOneOfs(proto);

    return t.exportNamedDeclaration(t.tsInterfaceDeclaration(
        t.identifier(name),
        null,
        [],
        t.tsInterfaceBody(
            Object.keys(proto.fields).map(fieldName => {
                const isOneOf = oneOfs.includes(fieldName);
                const field = proto.fields[fieldName];
                return tsPropertySignature(
                    t.identifier(fieldName),
                    t.tsTypeAnnotation(
                        getProtoField(field)
                    ),
                    getFieldOptionality(field, isOneOf)
                )
            })
        )
    ))
};


export const createCreateProtoType = (name: string, proto: ProtoType) => {
    const fields = Object.keys(proto.fields).map(key => {
        return {
            name: key,
            ...proto.fields[key]
        };
    })
        .map(field => {
            return t.objectProperty(
                t.identifier(field.name),
                getDefaultTSTypeFromProtoType(field)
            )
        })


    return functionDeclaration(t.identifier(getBaseCreateTypeFuncName(name)),
        [],
        t.blockStatement([
            t.returnStatement(t.objectExpression(
                [
                    ...fields,
                ]
            ))
        ]), false, false, t.tsTypeAnnotation(
            t.tsTypeReference(t.identifier(name))
        ))
};

