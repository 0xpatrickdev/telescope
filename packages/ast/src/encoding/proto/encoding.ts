import * as t from '@babel/types';
import { ProtoType } from './types';
import { fromPartialMethod } from './from-partial';
import { decodeMethod } from './decode';
import { encodeMethod } from './encode';
import { fromJSONMethod } from './from-json';
import { toJSONMethod } from './to-json';

export const createProtoObjectWithMethods = (name: string, proto: ProtoType) => {

    const methods = [
        encodeMethod(name, proto),
        decodeMethod(name, proto),
        fromJSONMethod(name, proto),
        toJSONMethod(name, proto),
        fromPartialMethod(name, proto),
    ];

    return t.exportNamedDeclaration(
        t.variableDeclaration('const',
            [t.variableDeclarator(
                t.identifier(name),
                t.objectExpression(
                    methods
                )
            )]
        )
    )
};