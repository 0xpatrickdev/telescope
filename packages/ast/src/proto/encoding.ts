import * as t from '@babel/types';
import { camel, pascal } from 'case';
import { identifier, tsEnumMember, tsPropertySignature, functionDeclaration, objectMethod } from '../utils';

import { ProtoType } from './types';

const protoEncodeMethod = (name: string, proto: ProtoType) => {
    return objectMethod(
        'method',
        t.identifier('encode'),
        [
            // args

            identifier('message', t.tsTypeAnnotation(
                t.tsTypeReference(
                    t.identifier('MsgJoinPool')
                )
            ), false),

            t.assignmentPattern(
                t.identifier('writer'),
                t.callExpression(
                    t.memberExpression(
                        t.memberExpression(
                            t.identifier('_m0'),
                            t.identifier('Writer')
                        ),
                        t.identifier('create')
                    ),
                    [

                    ]
                )
            )
        ],

        // body 
        t.blockStatement([

            /*
                if (message.sender !== "") {
                    writer.uint32(10).string(message.sender);
                }
            */
            t.ifStatement(
                t.binaryExpression('!==',
                    t.memberExpression(
                        t.identifier('message'),
                        t.identifier('sender')
                    ),
                    t.stringLiteral('')
                ),
                t.blockStatement([
                    t.expressionStatement(
                        t.callExpression(
                            t.memberExpression(
                                t.callExpression(
                                    t.memberExpression(
                                        t.identifier('writer'),
                                        t.identifier('uint32')
                                    ),
                                    [
                                        t.numericLiteral(10)
                                    ]
                                ),
                                t.identifier('string')
                            ),
                            [
                                t.memberExpression(
                                    t.identifier('message'),
                                    t.identifier('sender')
                                )
                            ]
                        )
                    )
                ])
            ),
            /*
            if (!message.poolId.isZero()) {
                writer.uint32(16).uint64(message.poolId);
            }
            */
            t.ifStatement(
                t.unaryExpression('!',
                    t.callExpression(
                        t.memberExpression(
                            t.memberExpression(
                                t.identifier('message'),
                                t.identifier('poolId')
                            ),
                            t.identifier('isZero')
                        ),
                        []
                    )
                ),
                t.blockStatement([
                    t.expressionStatement(
                        t.callExpression(
                            t.memberExpression(
                                t.callExpression(
                                    t.memberExpression(
                                        t.identifier('writer'),
                                        t.identifier('uint32')
                                    ),
                                    [
                                        t.numericLiteral(16)
                                    ]
                                ),
                                t.identifier('uint64')
                            ),
                            [
                                t.memberExpression(
                                    t.identifier('message'),
                                    t.identifier('poolId')
                                )
                            ]
                        )
                    )
                ])
            ),

            /*
            REPEAT OF FIRST PROP
            if (message.shareOutAmount !== "") {
                writer.uint32(26).string(message.shareOutAmount);
            }
            */

            t.ifStatement(
                t.binaryExpression(
                    '!==',
                    t.memberExpression(
                        t.identifier('message'),
                        t.identifier('shareOutAmount')
                    ),
                    t.stringLiteral('')
                ),
                t.blockStatement([
                    t.expressionStatement(
                        t.callExpression(
                            t.memberExpression(
                                t.callExpression(
                                    t.memberExpression(
                                        t.identifier('writer'),
                                        t.identifier('uint32')
                                    ),
                                    [
                                        t.numericLiteral(26)
                                    ]
                                ),
                                t.identifier('string')
                            ),
                            [
                                t.memberExpression(
                                    t.identifier('message'),
                                    t.identifier('shareOutAmount')
                                )
                            ]
                        )
                    )
                ])
            ),

            /*
                LOOP (ARRAY)
            */

            t.forOfStatement(
                t.variableDeclaration('const',
                    [
                        t.variableDeclarator(
                            t.identifier('v'),
                            null
                        )
                    ]
                ),
                t.memberExpression(
                    t.identifier('message'),
                    t.identifier('tokenInMaxs')
                ),
                t.blockStatement(
                    [
                        t.expressionStatement(
                            t.callExpression(
                                t.memberExpression(
                                    t.callExpression(
                                        t.memberExpression(
                                            t.identifier('Coin'),
                                            t.identifier('encode')
                                        ),
                                        [
                                            // "v!" just means it's NOT NULLABLE
                                            t.tsNonNullExpression(
                                                t.identifier('v')
                                            ),
                                            t.callExpression(
                                                t.memberExpression(
                                                    t.callExpression(
                                                        t.memberExpression(
                                                            t.identifier('writer'),
                                                            t.identifier('uint32')
                                                        ),
                                                        [
                                                            t.numericLiteral(34)
                                                        ]
                                                    ),
                                                    t.identifier('fork')
                                                ),
                                                []
                                            )
                                        ]
                                    ),
                                    t.identifier('ldelim')
                                ),
                                []
                            )
                        )
                    ]
                )
            ),

            /* RETURN writer */

            t.returnStatement(
                t.identifier('writer')
            )

        ]),
        false,
        false,
        false,
        // return type
        t.tsTypeAnnotation(
            t.tsTypeReference(
                t.tsQualifiedName(
                    t.identifier('_m0'),
                    t.identifier('Writer')
                )
            )
        )
    )
};

const protoDecodeMethod = (name: string, proto: ProtoType) => {
    return objectMethod(
        'method',
        t.identifier('decode'),
        [
            identifier('input',
                t.tsTypeAnnotation(
                    t.tsUnionType(
                        [
                            t.tsTypeReference(
                                t.tsQualifiedName(
                                    t.identifier('_m0'),
                                    t.identifier('Reader')
                                ),
                                null
                            ),
                            t.tsTypeReference(
                                t.identifier('Uint8Array')
                            )
                        ]
                    )
                ),
                false
            )
        ],
        t.blockStatement([

            /*
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
            */
            t.variableDeclaration(
                'const',
                [
                    t.variableDeclarator(
                        t.identifier('reader'),
                        t.conditionalExpression(
                            t.binaryExpression(
                                'instanceof',
                                t.identifier('input'),
                                t.memberExpression(
                                    t.identifier('_m0'),
                                    t.identifier('Reader')
                                )
                            ),
                            t.identifier('input'),
                            t.newExpression(
                                t.memberExpression(
                                    t.identifier('_m0'),
                                    t.identifier('Reader')
                                ),
                                [
                                    t.identifier('input')
                                ]
                            )
                        )
                    )
                ]
            ),

            /*
    let end = length === undefined ? reader.len : reader.pos + length;
            */

            t.variableDeclaration(
                'let',
                [
                    t.variableDeclarator(
                        t.identifier('end'),
                        t.conditionalExpression(
                            t.binaryExpression('===',
                                t.identifier('length'),
                                t.identifier('undefined')
                            ),
                            t.memberExpression(
                                t.identifier('reader'),
                                t.identifier('len')
                            ),
                            t.binaryExpression(
                                '+',
                                t.memberExpression(
                                    t.identifier('reader'),
                                    t.identifier('pos')
                                ),
                                t.identifier('length')
                            )
                        )
                    )
                ]
            ),

            /*
            
    const message = createBaseMsgJoinPool();

            */

            t.variableDeclaration(
                'const',
                [
                    t.variableDeclarator(
                        t.identifier('message'),
                        t.callExpression(
                            t.identifier('createBaseMsgJoinPool'),
                            []
                        )
                    )
                ]
            ),

            ///////////
            ///////////
            ///////////

            t.whileStatement(
                t.binaryExpression(
                    '<',
                    t.memberExpression(
                        t.identifier('reader'),
                        t.identifier('pos')
                    ),
                    t.identifier('end')
                ),
                t.blockStatement([

                    /// DECODE BODY
                    t.variableDeclaration(
                        'const',
                        [
                            t.variableDeclarator(
                                t.identifier('tag'),
                                t.callExpression(
                                    t.memberExpression(
                                        t.identifier('reader'),
                                        t.identifier('uint32')
                                    ),
                                    []
                                )
                            )
                        ]
                    ),

                    /*
           message.sender = reader.string();
              break;
                    */


                    t.switchStatement(
                        t.binaryExpression(
                            '>>>',
                            t.identifier('tag'),
                            t.numericLiteral(3)
                        ),
                        [
                            //
                            t.switchCase(
                                t.numericLiteral(1),
                                [
                                    t.expressionStatement(
                                        t.assignmentExpression(
                                            '=',
                                            t.memberExpression(
                                                t.identifier('message'),
                                                t.identifier('sender')
                                            ),
                                            t.callExpression(
                                                t.memberExpression(
                                                    t.identifier('reader'),
                                                    t.identifier('string')
                                                ),
                                                []
                                            )
                                        )
                                    ),
                                    t.breakStatement()
                                ]
                            ),

                            /*
                               case 2:
                                          message.poolId = (reader.uint64() as Long);
                                          break;
                            */

                            t.switchCase(
                                t.numericLiteral(2),
                                [
                                    t.expressionStatement(
                                        t.assignmentExpression(
                                            '=',
                                            t.memberExpression(
                                                t.identifier('message'),
                                                t.identifier('poolId')
                                            ),
                                            t.tsAsExpression(
                                                t.callExpression(
                                                    t.memberExpression(
                                                        t.identifier('reader'),
                                                        t.identifier('uint64')
                                                    ),
                                                    []
                                                ),
                                                t.tsTypeReference(
                                                    t.identifier('Long')
                                                )
                                            )
                                        )
                                    ),
                                    t.breakStatement()
                                ]
                            ),


                            /*
                             case 3:
                                          message.shareOutAmount = reader.string();
                                          break
                            */

                            t.switchCase(
                                t.numericLiteral(3),
                                [
                                    t.expressionStatement(
                                        t.assignmentExpression(
                                            '=',
                                            t.memberExpression(
                                                t.identifier('message'),
                                                t.identifier('shareOutAmount')
                                            ),
                                            t.callExpression(
                                                t.memberExpression(
                                                    t.identifier('reader'),
                                                    t.identifier('string')
                                                ),
                                                []
                                            )
                                        )
                                    ),
                                    t.breakStatement()
                                ]
                            ),

                            /*
                             case 4:
                                          message.tokenInMaxs.push(Coin.decode(reader, reader.unint32()));
                                          break;
                            */

                            t.switchCase(
                                t.numericLiteral(4),
                                [
                                    t.expressionStatement(
                                        t.callExpression(
                                            t.memberExpression(
                                                t.memberExpression(
                                                    t.identifier('message'),
                                                    t.identifier('tokenInMaxs')
                                                ),
                                                t.identifier('push')
                                            ),
                                            [
                                                t.callExpression(
                                                    t.memberExpression(
                                                        t.identifier('Coin'),
                                                        t.identifier('decode')
                                                    ),
                                                    [
                                                        t.identifier('reader'),
                                                        t.callExpression(
                                                            t.memberExpression(
                                                                t.identifier('reader'),
                                                                t.identifier('unint32')
                                                            ),
                                                            []
                                                        )
                                                    ]
                                                )
                                            ]
                                        )
                                    ),
                                    t.breakStatement()
                                ]
                            ),

                            /*
                            default:
                                    reader.skipType(tag & 7);
                                    break;
                            */

                            t.switchCase(
                                null,
                                [
                                    t.expressionStatement(
                                        t.callExpression(
                                            t.memberExpression(
                                                t.identifier('reader'),
                                                t.identifier('skipType')
                                            ),
                                            [
                                                t.binaryExpression(
                                                    '&',
                                                    t.identifier('tag'),
                                                    t.numericLiteral(7)
                                                )
                                            ]
                                        )
                                    ),
                                    t.breakStatement()
                                ]
                            )
                        ]
                    )

                ])
            ),

            // RETURN STATEMENT

            t.returnStatement(
                t.identifier('message')
            )

        ]),
        false,
        false,
        false,
        t.tsTypeAnnotation(
            t.tsTypeReference(
                t.identifier('MsgJoinPool')
            )
        )
    )
};

export const createProtoObjectWithMethods = (name: string, proto: ProtoType) => {
    return t.exportNamedDeclaration(
        t.variableDeclaration('const',
            [t.variableDeclarator(
                t.identifier('MsgJoinPool'),
                t.objectExpression(
                    [
                        protoEncodeMethod(name, proto),
                        protoDecodeMethod(name, proto)
                    ]
                )
            )]
        )
    )
};