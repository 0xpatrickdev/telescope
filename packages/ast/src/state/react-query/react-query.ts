
import * as t from '@babel/types';
import { callExpression, identifier, objectMethod, objectPattern } from '../../utils';
import { ProtoService, ProtoServiceMethod } from '@osmonauts/types';
import { GenericParseContext } from '../../encoding';
import { camel } from '@osmonauts/utils';
import { getRpcClassName } from '../../clients/rpc/class';
import { pascal } from 'case';

const makeUseHookName = (name: string) => {
  return camel('use_' + name);
};

const makeUseHookTypeName = (name: string) => {
  return pascal('Use_' + name + 'Query');
};

const makeHookKeyName = (name: string) => {
  return camel(name + 'Query');
};

const rpcHookMethod = (
  context: GenericParseContext,
  name: string,
  svc: ProtoServiceMethod
) => {
  const requestType = svc.requestType;
  const responseType = svc.responseType;
  const fieldNames = Object.keys(svc.fields ?? {})
  const hasParams = fieldNames.length > 0;

  let optional = false;
  // // if no params, then let's default to empty object for cleaner API
  if (!hasParams) {
    optional = true;
  } else if (hasParams && fieldNames.length === 1 && fieldNames.includes('pagination')) {
    // if only argument "required" is pagination
    // also default to empty
    optional = true;
  }

  // const methodArgs: t.Identifier = identifier(
  //     'request',
  //     t.tsTypeAnnotation(
  //         t.tsTypeReference(
  //             t.identifier(requestType)
  //         )
  //     ),
  //     optional
  // );


  context.addUtil('useQuery');

  return t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(makeUseHookName(name)),
      t.arrowFunctionExpression([
        objectPattern([
          t.objectProperty(
            t.identifier('request'),
            t.identifier('request'),
            false,
            true
          ),
          t.objectProperty(
            t.identifier('options'),
            t.identifier('options'),
            false,
            true
          )
        ],
          t.tsTypeAnnotation(
            t.tsTypeReference(
              t.identifier(makeUseHookTypeName(name)),
              t.tsTypeParameterInstantiation([
                t.tsTypeReference(
                  t.identifier('TData')
                )
              ])
            )
          )
        )
      ],
        t.blockStatement([
          t.returnStatement(
            callExpression(
              t.identifier('useQuery'),
              [
                t.arrayExpression([
                  t.stringLiteral(makeHookKeyName(name)),
                  t.identifier('request')
                ]),
                t.arrowFunctionExpression(
                  [],
                  t.blockStatement([
                    t.ifStatement(t.unaryExpression('!', t.identifier('queryService'), true), t.throwStatement(
                      t.newExpression(t.identifier('Error'), [
                        t.stringLiteral('Query Service not initialized')
                      ])
                    )),
                    t.returnStatement(
                      t.callExpression(t.memberExpression(t.identifier('queryService'), t.identifier(name)), [
                        t.identifier('request')
                      ])
                    )
                  ]))
              ],
              t.tsTypeParameterInstantiation([
                t.tsTypeReference(t.identifier(responseType)),
                t.tsTypeReference(t.identifier('Error')),
                t.tsTypeReference(t.identifier('Tdata'))
              ])
            ))
        ]))
    )
  ]);

  // return objectMethod('method', t.identifier(makeUseHookName(name)), [
  //     methodArgs
  // ], t.blockStatement([
  //     // t.returnStatement(
  //     //     t.callExpression(
  //     //         t.memberExpression(
  //     //             t.identifier('queryService'),
  //     //             t.identifier(name)
  //     //         ),
  //     //         [
  //     //             t.identifier('request')
  //     //         ]
  //     //     )
  //     // )
  // ]),
  //     false,
  //     false,
  //     false,
  //     null,
  //     t.tsTypeParameterDeclaration([
  //         t.tsTypeParameter(null, t.tsTypeReference(
  //             t.identifier(responseType)
  //         ), 'TData')
  //     ])
  // );
}

export const createRpcQueryHooks = (
  context: GenericParseContext,
  service: ProtoService
) => {

  context.addUtil('QueryClient');
  context.addUtil('createProtobufRpcClient');
  context.addUtil('ProtobufRpcClient');

  const camelRpcMethods = context.pluginValue('rpcClients.camelCase');
  // const name = service.name + 'ClientImpl';
  // const implementsName = service.name;
  // const methodNames = Object.keys(service.methods ?? {})
  //     .map(key => {
  //         return camelRpcMethods ? camel(key) : key
  //     });
  const methods = Object.keys(service.methods ?? {})
    .map(key => {
      const method = service.methods[key];
      const name = camelRpcMethods ? camel(key) : key;
      return rpcHookMethod(
        context,
        name,
        method
      )
    });

  const methodNames = Object.keys(service.methods ?? {})
    .map(key => {
      const name = camelRpcMethods ? camel(key) : key;
      return name;
    });

  return t.exportNamedDeclaration(
    t.variableDeclaration('const', [
      t.variableDeclarator(
        t.identifier('createRpcQueryHooks'),
        t.arrowFunctionExpression([
          identifier('rpc', t.tsTypeAnnotation(
            t.tsUnionType([
              t.tsTypeReference(t.identifier('ProtobufRpcClient')),
              t.tsUndefinedKeyword()
            ])
          ))],

          // body
          t.blockStatement([

            // query service
            t.variableDeclaration('const', [
              t.variableDeclarator(
                t.identifier('queryService'),
                t.callExpression(
                  t.identifier('getQueryService'),
                  [
                    t.identifier('rpc')
                  ]
                )
              )
            ]),


            ...methods,

            // return the methods...
            t.returnStatement(
              t.objectExpression(
                methodNames.map(name => t.objectProperty(
                  t.identifier(makeUseHookName(name)),
                  t.identifier(makeUseHookName(name)),
                  false,
                  true)
                )
              )
            )
          ])
          // end body

        )
      )
    ])
  );
};

