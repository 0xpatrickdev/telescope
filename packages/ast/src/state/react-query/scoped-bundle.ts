import * as t from '@babel/types';
import { GenericParseContext } from '../../encoding';
import { objectPattern } from '../../utils';

export const rpcHookFuncArguments = (): t.ObjectPattern[] => {
    return [
        objectPattern(
            [
                t.objectProperty(
                    t.identifier('rpc'),
                    t.identifier('rpc'),
                    false,
                    true
                )
            ],
            t.tsTypeAnnotation(
                t.tsTypeLiteral(
                    [
                        t.tsPropertySignature(
                            t.identifier('rpc'),
                            t.tsTypeAnnotation(
                                t.tsTypeReference(
                                    t.identifier('Rpc')
                                )
                            )
                        )
                    ]
                )
            )
        )
    ];
};

export const rpcHookClassArguments = (): t.ObjectExpression[] => {
    return [
        t.objectExpression(
            [
                t.objectProperty(
                    t.identifier('rpc'),
                    t.identifier('rpc'),
                    false,
                    true
                )
            ]
        )
    ];
};

export const rpcHookNewTmRequire = (
    path: string,
    className: string
) => {
    return t.callExpression(
        t.memberExpression(
            t.callExpression(
                t.identifier('require'),
                [
                    t.stringLiteral(
                        path
                    )
                ]
            ),
            t.identifier(className),
            false
        ),
        [
            t.identifier('rpc')
        ]
    )

}

export const rpcHookRecursiveObjectProps = (
    names: string[],
    leaf?: any
) => {
    const [name, ...rest] = names;

    let baseComponent;
    if (names.length === 1) {
        baseComponent = leaf ? leaf : t.identifier(name)
    } else {
        baseComponent = rpcHookRecursiveObjectProps(rest, leaf)
    }

    return t.objectExpression([
        t.objectProperty(
            t.identifier(name),
            baseComponent
        )
    ])
};

export const rpcHookTmNestedImportObject = (
    obj: object,
    className: string
) => {

    if (typeof obj === 'string') {
        return rpcHookNewTmRequire(obj, className);
    }

    const keys = Object.keys(obj);

    return t.objectExpression(keys.map(name => {
        return t.objectProperty(
            t.identifier(name),
            rpcHookTmNestedImportObject(obj[name], className)
        )
    }))
};

export const createScopedRpcHookFactory = (
    context: GenericParseContext,
    obj: object,
    identifier: string
) => {

    context.addUtil('ProtobufRpcClient');

    return t.exportNamedDeclaration(
        t.variableDeclaration(
            'const',
            [
                t.variableDeclarator(
                    // createRPCQueryHooks
                    t.identifier(identifier),
                    t.arrowFunctionExpression(
                        [
                            objectPattern([
                                t.objectProperty(
                                    t.identifier('rpc'),
                                    t.identifier('rpc'),
                                    false,
                                    true
                                )
                            ], t.tsTypeAnnotation(
                                t.tsTypeLiteral([
                                    t.tsPropertySignature(
                                        t.identifier('rpc'),
                                        t.tsTypeAnnotation(
                                            t.tsUnionType([
                                                t.tsTypeReference(
                                                    t.identifier('ProtobufRpcClient')
                                                ),
                                                t.tsUndefinedKeyword()
                                            ])
                                        )
                                    )
                                ])
                            ))
                        ],
                        t.blockStatement([

                            t.returnStatement(
                                rpcHookTmNestedImportObject(
                                    obj,
                                    'createRpcQueryHooks'
                                )
                            )

                        ]),
                        false
                    )
                )
            ]
        )
    )
}
