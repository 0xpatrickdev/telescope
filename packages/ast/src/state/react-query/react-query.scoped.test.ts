import { traverse, getNestedProto } from '@osmonauts/proto-parser'
import { defaultTelescopeOptions, ProtoService } from '@osmonauts/types';
import { expectCode, getGenericParseContext, getTestProtoStore } from '../../../test-utils';
import { GenericParseContext } from '../../encoding';
import { createRpcQueryHooks, createRpcQueryHookInterfaces } from './react-query';
import * as t from '@babel/types';
import { createScopedRpcHookFactory } from './scoped-bundle';

const store = getTestProtoStore();
store.traverseAll();

it('createScopedRpcHookFactory', async () => {
    const context = getGenericParseContext();
    expectCode(
        t.program(
            createScopedRpcHookFactory(
                context,
                {
                    cosmos: {
                        bank: {
                            v1beta1: "./proto/cosmos/bank/v1beta1/a.lcd"
                        },
                        gov: {
                            v1beta1: "./proto/cosmos/bank/v1beta1/b.lcd"
                        },
                    },
                    osmosis: {
                        gamm: {
                            v1beta1: "./proto/cosmos/bank/v1beta1/c.lcd"
                        }
                    }
                },
                'createRpcFactorizzle',
            )
        )
    )
});
