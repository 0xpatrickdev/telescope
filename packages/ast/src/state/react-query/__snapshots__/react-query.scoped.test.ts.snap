// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createScopedRpcHookFactory 1`] = `
"export const createRpcFactorizzle = ({
  rpc
}: {
  rpc: ProtobufRpcClient | undefined;
}) => {
  return {
    cosmos: {
      bank: {
        v1beta1: require("./proto/cosmos/bank/v1beta1/a.lcd").createRpcQueryHooks(rpc)
      },
      gov: {
        v1beta1: require("./proto/cosmos/bank/v1beta1/b.lcd").createRpcQueryHooks(rpc)
      }
    },
    osmosis: {
      gamm: {
        v1beta1: require("./proto/cosmos/bank/v1beta1/c.lcd").createRpcQueryHooks(rpc)
      }
    }
  };
};"
`;
